white-paper/wp.tex

```diff
@@ -664,7 +664,24 @@ \subsection{Features}
     nodes, e.g.\@ how many nodes contains each repository, the state of nodes, etc.
 \end{description}
 
-\section{Compare to other solutions}
+\subsection{Load balancing}\label{sec:load-balancing}
+
+The front-end decides how to redirect incoming fetch requests to replica nodes.
+It knows network addresses and current availability of repository replicas,
+so it can decide how to route next fetch request to git repository back-end storage node.
+On read access to repository, it sends a health check message to replica nodes to be sure
+which are available. It stores the result in local cache with configurable time to live
+properties. When front-end unable to access some of back-end node, it marks it as unavailable
+and forget about it for some time. Then in check availability periodically.
```

Comment:
Should front-end checks health of nodes by itself. Or nodes could send heartbeats every T seconds and front end only listen to?
In future, when we will need in self healing mechanism for repos, where 1 or 2 nodes have failed, who will cares about it and takes decision to start new nodes?  If front end will, It is needed to realize that some node has failed a bit earlier, than a request to a repo came. @g4s8  WDYT?

Reply:
@stepanvalyavskiy in that case front-end node should listen for all back-end nodes health, we can't rely on UDP broadcasts, since nodes could be located in different regions. So the network will be full of health-check TCP connections from each back-end to each front-end (since each front-end may access each back-end). Assuming if we have about 5000 back-end nodes and 500 front-end nodes, we'll have `2500000` health-check connections in a system to keep the information up to date